### ISSUE TITLE
net: support switching between ethernet and wireless

### ISSUE THREAD

**Issue by mvdan** [greshake/i3status-rust#891]
This block works fine for my wireless device:

```
[[block]]
block = "net"
device = "wlan0"
speed_up = true
interval = 2
```

However, when I connect ethernet and disable wifi, as expected the block then does nothing. It would be nice to instead see the traffic for the wired device, in my case `ens8u2`.

Here are two ideas to potentially solve this:

1) Default to using the device for the first default route in `ip route`; most setups will only have one such device. If the wrong one is detected, the user can always specify one via `device = "..."`.

2) Allow using a list of fallback devices, where the first one with non-zero traffic is used. In my case, I could use something like `devices = ["wlan0", "ens8u2"]`.

Both options would require extra work, but I see no alternative if we want to automatically switch over between two devices.

**Comment by ammgws** [comment]
How about defining two blocks in your config and using `allow_missing` and `hide_missing`?

**Comment by mvdan** [comment]
I thought about that, but that takes up more space. Right now, the "empty" wlan0 net block contains two icons: the wireless signal icon, and an "x" presumably to say that the device isn't enabled.

**Comment by mvdan** [comment]
Perhaps option 3 is to make the block be omitted when the input (device) is disabled.

**Comment by ammgws** [comment]
IIRC isn't that what `hide_missing` does? (hiding the block when the device can't be found)

**Comment by mvdan** [comment]
Perhaps. I have to admit that the docs don't clarify the difference between missing and inactive, and the code in `net.rs` seems to use both in the same way as far as I can see.

I tried both, and I still see the wlan0 widget:

```
$ grep -C3 wlan0 .config/i3status-rust/config.toml 

[[block]]
block = "net"
device = "wlan0"
speed_up = true
interval = 2
hide_inactive = true
$ ip l | grep wlan0
4: wlan0: <BROADCAST,MULTICAST> mtu 1500 qdisc noqueue state DOWN mode DORMANT group default qlen 1000
```

I completely reloaded sway after changing the config each time, so that fully restarts i3status-rust. I did add the wired net block, which now shows up.

If this worked I'd be fine with it :)

**Comment by GladOSkar** [comment]
I actually use both `hide_missing = true` + `hide_inactive = true` on each interface for your exact purpose, only showing those interfaces that i'm connected to.

Only issue is with alternating color schemes (see #349) but that's less important

**Comment by ammgws** [comment]
@mvdan You need both `hide_inactive` and `hide_missing` in your config for it to hide the block

**Comment by mvdan** [comment]
Oh, that's not what I gathered from https://github.com/greshake/i3status-rust/blob/master/blocks.md#net then. Why would they be separate options, if they appear to do nothing on their own?

**Comment by GladOSkar** [comment]
It's true, i found this confusing as well. IMO it should mean:

`missing` implies that the interface has been disabled entirely. Think a VPN that is not running.

`inactive` implies that the interface is available, but not connected. Think wifi not found / ethernet not connected.

But in reality it seems neither work on their own - both my disabled and my disconnected interfaces are only hidden when both options are `true`. Is this a bug @ammgws?

**Comment by ammgws** [comment]
Can't look into it atm but they were added in these PRs in the order below. Perhaps reading through the discussions may shed some light on the intended behaviour if the code itself isn't clear enough:

This PR added `hide_inactive`: https://github.com/greshake/i3status-rust/pull/102

This PR added `hide_missing`: https://github.com/greshake/i3status-rust/pull/344

This one added documentation for `hide_missing` and `hide_inactive`: https://github.com/greshake/i3status-rust/pull/476

**Comment by GladOSkar** [comment]
Interesting. The blurb for #102 says

> Automatic hiding of graphs/throughput info for inactive networks (which can be turned off with `hide_inactive = false`)

so that was initially only intended to hide the graph, not the block. But judging by the code it appears to have hidden the entire block anyway:

https://github.com/atheriel/i3status-rust/blob/bdd359fdebba223a336cb7a368db442d4d109a15/src/blocks/net.rs#L397-L401

PR #344 then was probably the culprit: This code looks like it was intended to be `hide_missing` OR `hide_hidden`, but because both are inverted it's actually a `hide_missing` AND `hide_hidden` for the `else` clause:

https://github.com/greshake/i3status-rust/blob/2bb498fba6e4cae9a2aa7b5b6856f27999d96167/src/blocks/net.rs#L512-L516

**Comment by ammgws** [comment]
Looks like that first one should have been
```rust
if self.active {
...
 } else if self.hide_inactive { 
     vec![&self.network] 
 } else { 
     vec![] 
 } 
```
in order for it to only show the basic info widget and not the graph widgets.

Although now those widgets have been replaced by the format string..

**Comment by GladOSkar** [comment]
No the first one was correct i think. I'm currently debugging and will make a PR right after

**Comment by GladOSkar** [comment]
Alright it's in #897, feel free to have a look
